mkdocs/commands/serve.py

```diff
@@ -52,7 +55,12 @@ def get_config():
     config = get_config()
     config.plugins.on_startup(command=('build' if is_clean else 'serve'), dirty=is_dirty)
 
-    host, port = config.dev_addr
+    host = config.dev_addr.host
+    if port is None:
+        port = config.dev_addr.port
+    if port is None:
+        port = get_random_port(config)
```

Comment:
Does it mean that it won't always start with 8000 anymore? It would be a strong -1 for me.

Reply:
What do you mean by "start"?

Reply:
I mean that if I currently have zero server instance running, I expect `mkdocs serve` to run on port 8000. Then only if I run additional instances, I expect it to use random ports.

Reply:
Yeah it will not simply pick 8000 anymore. I describe why I think it should be done that way in the top post. To get the old behavior you can pass `-p 8000`

Reply:
That will break a lot of existing resources. And not knowing which port will be used each time I run `mkdocs serve` in a different project will be super annoying IMO. I recommend not doing that and keeping 8000 as default. It would probably make things more complex (catch the error that the server could not run because the port is already used, and try again on 3 random ports).

Reply:
I don't think that starting with exactly 8000 is more complex, I think it's *worse*. But let's keep discussing and mentioning actual usecases that can be broken

Reply:
Might be technically worse, but that's what `mkdocs serve` always did, so changing it would need a deprecation period at least.

Reply:
Well imagine if Flask or Uvicorn suddenly started serving stuff on a different port only *sometimes* 😕 but they don't

We don't need to move forward in my way but I don't think this is the right way either

Reply:
If you say it's so important that the content is served at port 8000 and scripts rely on it, wouldn't we be introducing silent bugs into scripts - mkdocs serve successfully starts but actually some other content is served on port 8000

Reply:
Amazing!

Reply:
As per my prior messages, I think trying several ports is maybe not a good thing.

Instead I want to encourage users to pick a distinct port for each of their sites.

Here is the alternative I propose:
* https://github.com/mkdocs/mkdocs/pull/3498/commits/3dd617b1e6fc6b1e9efa9c2d6de6c0bc10eb7ad8
* https://github.com/mkdocs/mkdocs/pull/3498#issuecomment-1837268183

Reply:
I absolutely second that we should keep 8000 as a default. Please keep in mind that MkDocs is used by many not particularly tech-savvy users. Allowing for just that is a critical factor of its success.

Reply:
It is seems reasonable to keep other things as much as much as there were as possible, typically to keep the original 8000 port as default (principle of least astonishment); on the other hand, flexibility is always welcome.

Reply:
An other argument would be: look at what others do. I mentioned Flask and Uvicorn earlier: they do not randomize ports. If the default or specified port is in use, it errors out, as the user must specify another one, simple as that. If anyone knows a tool that retries with randomized ports, please share!